src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibraryHelper.java

```diff
@@ -1367,6 +1368,72 @@ private CppCompilationContext initializeCppCompilationContext(CppModel model) {
+      internalHeaderMapInfo.addHeaders(publicTextualHeaders);
+      Artifact internalHeaderMap = ruleContext.getPackageRelativeArtifact(PathFragment.create(targetName + "_internal.hmap"),
+                ruleContext
+                .getConfiguration()
+                .getGenfilesDirectory(ruleContext.getRule().getRepository()));
+      ruleContext.registerAction(
+        new HeaderMapAction(ruleContext.getActionOwner(),
+            internalHeaderMapInfo.build().getSources(),
+            internalHeaderMap));
+      contextBuilder.addQuoteIncludeDir(internalHeaderMap.getExecPath());
+      headerMapsBuilder.add(internalHeaderMap);
+
+      String namespace;
+      if (ruleContext.attributes().has("header_namespace")) {
+         namespace = ruleContext.attributes().get("header_namespace", Type.STRING);
+      } else {
+         namespace = ruleContext.getRule().getName();
+      }
+
+      // Construct the dep headermap.
+      // This header map additionally contains namespaced headers so that a user
+      // can import headers of the form Namespace/Header.h from headers within
+      // the current target.
+      HeaderMapInfo.Builder depHeaderMapInfo = new HeaderMapInfo.Builder();
+      depHeaderMapInfo.setNamespace(namespace);
+      depHeaderMapInfo.addNamespacedHeaders(publicHeaders.getHeaders());
+      depHeaderMapInfo.addHeaders(publicHeaders.getHeaders());
+      depHeaderMapInfo.addNamespacedHeaders(privateHeaders);
+      depHeaderMapInfo.addHeaders(privateHeaders);
+      depHeaderMapInfo.addNamespacedHeaders(publicTextualHeaders);
+      depHeaderMapInfo.addHeaders(publicTextualHeaders);
+
+      // Merge all of the header map info from deps. The headers within a given
+      // target have precedence over over dep headers ( See
+      // HeaderMapInfo.build() ).
+      if (ruleContext.attributes().has("deps")){
+        for (HeaderMapInfoProvider hmapProvider : ruleContext.getPrerequisites("deps", Mode.TARGET, HeaderMapInfoProvider.class)) {
+          depHeaderMapInfo.mergeHeaderMapInfo(hmapProvider.getInfo());
+        }
+      }
+      Artifact depHeaderMap = ruleContext.getPackageRelativeArtifact(PathFragment.create(targetName + ".hmap"),
+                ruleContext
+                .getConfiguration()
+                .getGenfilesDirectory(ruleContext.getRule().getRepository()));
+      ruleContext.registerAction(
+        new HeaderMapAction(ruleContext.getActionOwner(),
+            depHeaderMapInfo.build().getSources(),
+            depHeaderMap));
+      contextBuilder.addIncludeDir(depHeaderMap.getExecPath());
+      headerMapsBuilder.add(depHeaderMap);
+
+      // If we have headermaps, then we need to add an include of
+      // the working directory ( i.e. exec root ) in this form
+      // and it must be after including the header map files
+      contextBuilder.addIncludeDir(PathFragment.create("."));
+	    ImmutableList headerMaps = headerMapsBuilder.build();
```

Comment:
SuperNit: Reformat pls. In general, Bazel follows google java style guide (https://google.github.io/styleguide/javaguide.html). I'll take care of that when importing, but to increase readability pls try to adhere to column limit: 100, extra points for https://google.github.io/styleguide/javaguide.html#s4.5.1-line-wrapping-where-to-break :) But I'm nitting.

Reply:
Cool, thanks for the tip. I'll try to make this follow to the style guide closely :)